### PR TITLE
Adds the extended keycode where appropriate and updates to the latest IronRDP master commit hash.

### PR DESCRIPTION
**Pull request by ibeckermayer**
Both of these changes are necessary in order to get the windows key to work again.

Corresponds with https://github.com/Devolutions/IronRDP/pull/364

changelog: Allow users to send the Windows key for desktop sessions.

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 1]
```diff
@@ -447,6 +464,11 @@ impl Client {
         if !key.down {
             flags = KeyboardFlags::RELEASE;
         }
+        let extended = key.code & 0xE000 == 0xE000;
+        if extended {
+            flags |= KeyboardFlags::EXTENDED;
+        }
```
Comment (ibeckermayer) [inline lib/srv/desktop/rdp/rdpclient/src/client.rs:470]: This is where we actually set the `EXTENDED` flag; the rest of the changes are to interface with the updated IronRDP apis.

**Review by zmb3 — APPROVED** [review]
Seems fine to me. Let's get the disconnect reason plumbed through to the UI soon though. That was a really nice enhancement when we added it for rdp-rs

**Review by probakowski — APPROVED** [review]

**Comment by ibeckermayer** [comment]
@zmb3 @probakowski Please take another look.

I ended up needing to include some pretty substantive changes in order to pipe messages through to the client itself. I also realize we forgot to swap in the desktopplayback endpoint in https://github.com/gravitational/teleport/pull/37962 new authenticated websocket paradigm, so those changes are included now as well.

**Comment by probakowski** [comment]
Are changes around playback ws the same here as in https://github.com/gravitational/teleport/pull/37981? If so would it make sense to based this PR on top of that one?

File: web/packages/teleport/src/DesktopSession/DesktopSession.tsx [hunk 2]
```diff
@@ -58,92 +51,97 @@ declare global {
   }
 }
 
+const invalidStateMessage =
```
Comment (zmb3) [inline web/packages/teleport/src/DesktopSession/DesktopSession.tsx:54]: I don't know how I feel about having the product say "something went wrong, please file a bug report." Can we just display a more generic error and log whatever information we can?

File: web/packages/teleport/src/DesktopSession/DesktopSession.tsx [hunk 3]
```diff
@@ -58,92 +51,97 @@ declare global {
...
+        } else if (showAlert) {
+          let message = '';
+          if (fetchAttempt.status === 'failed') {
+            message = fetchAttempt.statusText || 'fetch attempt failed';
+          } else if (tdpConnection.status === 'failed') {
+            message = tdpConnection.statusText || 'TDP connection failed';
+          } else if (tdpConnection.status === '') {
+            message =
+              tdpConnection.statusText || 'TDP connection ended gracefully';
+          } else if (wsConnection.status === 'closed') {
```
Comment (zmb3) [inline web/packages/teleport/src/DesktopSession/DesktopSession.tsx:111]: The number of nested if/else blocks in here is 🤯

Any way we can simplify this?

File: web/packages/teleport/src/Player/DesktopPlayer.tsx [hunk 4]
```diff
@@ -175,6 +177,14 @@ const useDesktopPlayer = ({ clusterId, sid }) => {
     [setPlayerStatus, setStatusText]
   );
 
+  const clientOnTdpInfo = useCallback(
+    (info: string) => {
+      setPlayerStatus(StatusEnum.COMPLETE);
+      setStatusText(info);
+    },
+    [setStatusText]
```
Comment (zmb3) [inline web/packages/teleport/src/Player/DesktopPlayer.tsx:185]: Why do we include `setStatuxText` but not `setPlayerStatus`?

In any case, we can probably omit both. The React hooks docs say:
> React guarante
es that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

File: web/packages/teleport/src/lib/tdp/client.ts [hunk 5]
```diff
@@ -134,7 +136,11 @@ export default class Client extends EventEmitterWebAuthnSender {
     // prior to a socket 'close' event (https://stackoverflow.com/a/40084550/6277051).
     // Therefore, we can rely on our onclose handler to account for any websocket errors.
     this.socket.onerror = null;
-    this.socket.onclose = () => {
+    this.socket.onclose = ev => {
+      let message = 'session disconnected';
+      if (ev.code !== WebsocketCloseCode.NORMAL) {
+        message = `connection closed with websocket error code: ${ev.code}`;
```
Comment (zmb3) [inline web/packages/teleport/src/lib/tdp/client.ts:142]: Users aren't going to have any idea what various websocket error codes mean. This seems like something that could be logged but doesn't need to be user-facing.

File: web/packages/teleport/src/DesktopSession/DesktopSession.tsx (hunk 3 above)
Comment (ibeckermayer) [reply web/packages/teleport/src/DesktopSession/DesktopSession.tsx]: I did another minor overhaul to make the state machine more comprehensible: 6cfb564322a1deedff79e1451b2d5982d8f92bbe

File: web/packages/teleport/src/lib/tdp/client.ts (hunk 5 above)
Comment (ibeckermayer) [reply web/packages/teleport/src/lib/tdp/client.ts]: Updated in 6cfb564322a1deedff79e1451b2d5982d8f92bbe

File: web/packages/teleport/src/DesktopSession/DesktopSession.tsx (hunk 2 above)
Comment (ibeckermayer) [reply web/packages/teleport/src/DesktopSession/DesktopSession.tsx]: Changed to `const invalidStateMessage = 'internal application error';` in 6cfb564322a1deedff79e1451b2d5982d8f92bbe

File: web/packages/teleport/src/Player/DesktopPlayer.tsx (hunk 4 above)
Comment (ibeckermayer) [reply web/packages/teleport/src/Player/DesktopPlayer.tsx]: c4c4aa2c6e2cad87db660351b95cc66391bdaf46

File: lib/srv/desktop/rdp/rdpclient/client.go [hunk 6]
```diff
@@ -305,14 +310,37 @@ func (c *Client) startRustRDP(ctx context.Context) error {
...
+	if res.err_code != C.ErrCodeSuccess {
+		var err error
+
+		if message != "" {
+			err = trace.Errorf("RDP client exited with an error: %v", message)
+		} else {
+			err = trace.Errorf("RDP client exited with an unknown error")
+		}
+
+		c.sendTDPNotification(err.Error(), tdp.SeverityError)
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/client.go:331]: This is okay for now, but Rust errors are generally not suitable for display in the UI. Let's file an action item to clean this up.

File: lib/srv/desktop/rdp/rdpclient/client.go [hunk 7]
```diff
@@ -305,14 +310,37 @@ func (c *Client) startRustRDP(ctx context.Context) error {
...
+		return err
+	}
+
+	if message != "" {
+		message = fmt.Sprintf("RDP client exited gracefully with message: %v", message)
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/client.go:336]: This is better than what we had, but still not super user friendly:
image

With rdp-rs we had the lookup table with human readable messages for each disconnect code: 

https://github.com/gravitational/rdp-rs/blob/edfb5330a11d11eaf36d65e4300555368b4c6b02/src/core/global.rs#L309-L339

Doesn't need to be addressed in this PR but I think we should get it done soon.

**Review by zmb3 — APPROVED** [review]

File: lib/srv/desktop/rdp/rdpclient/client.go (hunk 7 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go:339]: https://github.com/gravitational/teleport/issues/38373

File: lib/srv/desktop/rdp/rdpclient/client.go (hunk 6 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go:334]: https://github.com/gravitational/teleport/issues/38373

**Comment by ibeckermayer** [comment]
> Are changes around playback ws the same here as in #37981? If so would it make sense to based this PR on top of that one?

That may have been the case when you asked the question, but everything should be up to date now.

**Review by probakowski — APPROVED** [review]

**Pull request merged**